### ISSUE TITLE
wrong used-before-assignment in python3.4

### ISSUE THREAD

**Issue by bjm5vpfkew** [PyCQA/pylint#853]
I'm running pylint in an ubuntu 14.04 virtualenv of python3.4

test.py as textfile: [test.txt](https://github.com/PyCQA/pylint/files/177520/test.txt)
`pip list` shows: [requirements.txt](https://github.com/PyCQA/pylint/files/177521/requirements.txt)

result is:

```
$ pylint test.py
No config file found, using default configuration
************* Module test
E:  8,23: Using variable 'message' before assignment (used-before-assignment)


Report
======
6 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |1      |NC         |NC         |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |0      |NC         |NC         |0           |0        |
+---------+-------+-----------+-----------+------------+---------+
|method   |0      |NC         |NC         |0           |0        |
+---------+-------+-----------+-----------+------------+---------+
|function |1      |NC         |NC         |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+



Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |9      |75.00 |NC       |NC         |
+----------+-------+------+---------+-----------+
|docstring |2      |16.67 |NC       |NC         |
+----------+-------+------+---------+-----------+
|comment   |0      |0.00  |NC       |NC         |
+----------+-------+------+---------+-----------+
|empty     |1      |8.33  |NC       |NC         |
+----------+-------+------+---------+-----------+



Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |NC       |NC         |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |NC       |NC         |
+-------------------------+------+---------+-----------+



Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |0      |NC       |NC         |
+-----------+-------+---------+-----------+
|refactor   |0      |NC       |NC         |
+-----------+-------+---------+-----------+
|warning    |0      |NC       |NC         |
+-----------+-------+---------+-----------+
|error      |1      |NC       |NC         |
+-----------+-------+---------+-----------+



Messages
--------

+-----------------------+------------+
|message id             |occurrences |
+=======================+============+
|used-before-assignment |1           |
+-----------------------+------------+



Global evaluation
-----------------
Your code has been rated at 1.67/10

```

This problem does not exist in python2.7

**Comment by The-Compiler** [comment]
I'm not sure there's much pylint can do here - after all, it might be right, if your `range()` would raise an exception. And it can't be certain there is no exception raised before you do `message = object()`.

**Comment by bjm5vpfkew** [comment]
If range raises, finally is called... message might not be set. in finally message = object() is set...
don't see the problem @The-Compiler

**Comment by bjm5vpfkew** [comment]
And it works in python2.7

**Comment by The-Compiler** [comment]
Err, sorry, ignore my comment - I somehow thought you were *using* `message` in `finally`, but you're setting it.

**Comment by PCManticore** [comment]
I can reproduce the issue, thanks for letting us know!

**Comment by kengruven** [comment]
The loop isn't necessary.  Here's an even simpler example that flags a `used-before-assignment` false positive:

```
try:
    MY_INT = 1
    print("MY_INT = %d" % MY_INT)
finally:
    MY_INT = 2
```

**Comment by ppolewicz** [comment]
A slightly different reproduction with empty `try` block:

code:
```python
try:
    pass
except:
    FALSE_POSITIVE = 1
    FALSE_POSITIVE  # here pylint claims used-before-assignment
finally:
    FALSE_POSITIVE = 2  # this line is needed to reproduce the issue
```

occurrence:
```
$ pylint lintcheck | grep FALSE
No config file found, using default configuration
E:  5, 4: Using variable 'FALSE_POSITIVE' before assignment (used-before-assignment)
(pylint3) rare@rare-VirtualBox ~/b2 $ 
```

environment:
```
$ pylint --version
No config file found, using default configuration
pylint 1.6.4, 
astroid 1.4.8
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4]
(pylint3) rare@rare-VirtualBox ~/b2 $ cat lintcheck.py
```

Does *not* reproduce on Python 2.7.
